source/extensions/http/header_validators/envoy_default/header_validator.cc

```diff
@@ -8,38 +14,344 @@ namespace EnvoyDefault {
+  //
+  // Verify that the header value is valid.
+  //
+  // From RFC 7230, https://datatracker.ietf.org/doc/html/rfc7230:
+  //
+  // header-field   = field-name ":" OWS field-value OWS
+  // field-value    = *( field-content / obs-fold )
+  // field-content  = field-vchar [ 1*( SP / HTAB ) field-vchar ]
+  // field-vchar    = VCHAR / obs-text
+  // obs-text       = %x80-FF
+  //
+  // VCHAR          =  %x21-7E
+  //                   ; visible (printing) characters
+  //
+  const auto& value_string_view = value.getStringView();
+  bool is_valid = true;
+
+  for (std::size_t i{0}; i < value_string_view.size() && is_valid; ++i) {
+    is_valid = testChar(kGenericHeaderValueCharTable, value_string_view.at(i));
+  }
+
+  if (!is_valid) {
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidCharacters};
+  }
+
   return HeaderEntryValidationResult::success();
 }
 
 ::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
-HeaderValidator::validateResponseHeaderEntry(const HeaderString&, const HeaderString&) {
+HeaderValidator::validateContentLengthHeader(const HeaderString& value) {
+  //
+  // From RFC 7230, https://datatracker.ietf.org/doc/html/rfc7230#section-3.3.2:
+  //
+  // Content-Length = 1*DIGIT
+  //
+  const auto& value_string_view = value.getStringView();
+
+  if (value_string_view.empty()) {
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidContentLength};
+  }
+
+  auto buffer_start = value_string_view.data();
+  auto buffer_end = buffer_start + value_string_view.size();
+
+  std::uint32_t int_value{};
+  auto result = std::from_chars(buffer_start, buffer_end, int_value);
+  if (result.ec == std::errc::invalid_argument || result.ptr != buffer_end) {
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidContentLength};
+  }
+
   return HeaderEntryValidationResult::success();
 }
 
-::Envoy::Http::HeaderValidator::RequestHeaderMapValidationResult
-HeaderValidator::validateRequestHeaderMap(::Envoy::Http::RequestHeaderMap&) {
-  return RequestHeaderMapValidationResult::success();
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+HeaderValidator::validateHostHeader(const HeaderString& value) {
+  //
+  // From RFC 7230, https://datatracker.ietf.org/doc/html/rfc7230#section-5.4,
+  // and RFC 3986, https://datatracker.ietf.org/doc/html/rfc3986#section-3.2.2:
+  //
+  // Host       = uri-host [ ":" port ]
+  // uri-host   = IP-literal / IPv4address / reg-name
+  //
+  const auto& value_string_view = value.getStringView();
+
+  auto user_info_delimiter = value_string_view.find('@');
+  if (user_info_delimiter != absl::string_view::npos) {
+    // :authority cannot contain user info, reject the header
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidHost};
+  }
+
+  // identify and validate the port, if present
+  auto port_delimiter = value_string_view.find(':');
+  auto host_string_view = value_string_view.substr(0, port_delimiter);
```

Comment:
It looks like validation of the `host` is missing. The current more stricter implementation distinguishes between IPv6address and reg-name, treating IPv4address as reg-name. For IPv6 it only checks that character set is HEX and `:` and also allows zone id, which the https://www.rfc-editor.org/rfc/rfc3986#section-3.2.2 does not specify.

So I think we can just check that host is either IPv6address consisting of HEX and ':', we do not need to check if it parses to a valid IPv6 address. And then check for reg-name, which should cover IPv4Address as well.

Reply:
We have these utilities https://github.com/envoyproxy/envoy/blob/fd17464986ce6113354dbb1154f6ef7b18ffbd92/source/common/network/utility.cc#L123 for parsing authority but they are a bit heavy for what we need, since they actually produce parsed IP address.

I think we need to bite the bulled and write validation. Right now we validate the following:

1. if host starts with '[', validate that it either ends with ']' or ']:port' and there hexdigit and ':' in between []. No validation that it is actually a valid IPv6 address otherwise.
2. check the `:port` at the end and validate port is uint16_t
3. check that all characters up to end or `:port` are unreserved / pct-encoded / sub-delims This covers both IPv4 and reg-name

Reply:
@ameily can this comment be taken care, so that https://github.com/envoyproxy/envoy/issues/22859 gets fixed as well?

Reply:
I added basic syntax checking for IPv6 addresses that only validates that allowable characters are between square brackets: `\[[A-Fa-f0-9:]+\]`. Do you think that the validation needs to go further and actually check each segment and total address length?

Reply:
Hi @ameily ,
From what I understand you check that inside the brackets you have only characters that are allowed.
What would happen if the ipv6 address was provided without any bracket (e.g: 2001:db8::2:1:443)?
Would that still fail validation?